File: apps/api/src/reviews/reviews.service.ts
import {
  BadRequestException,
  ConflictException,
  ForbiddenException,
  Injectable,
  NotFoundException,
} from "@nestjs/common";
import { PrismaService } from "../common/prisma/prisma.service";
import { AuthenticatedUser } from "../common/decorators/current-user.decorator";
import { CreateReviewDto } from "./dto/create-review.dto";
import { CreateReviewResponseDto } from "./dto/create-review-response.dto";

// The `reviews` Prisma model, the search service's `most_reviewed` /
// `highest_verified_rating` sort options (apps/api/src/search/search.service.ts),
// and the public property response (public.service.ts's getPropertyBySlug
// already `include`s `reviews: { include: { categories: true } }`) all
// existed already. Nothing ever wrote a review or exposed a way to respond
// to one — this module closes that gap without touching any of the above.
@Injectable()
export class ReviewsService {
  constructor(private readonly prisma: PrismaService) {}

  /**
   * Only the tenant on a COMPLETED booking may review the stay, and only
   * once per booking — this is the "verified stays only" rule referenced by
   * the Review model's own `bookingId` comment.
   */
  async createForBooking(user: AuthenticatedUser, bookingId: string, dto: CreateReviewDto) {
    const booking = await this.prisma.booking.findUnique({
      where: { id: bookingId },
      include: { unit: { include: { property: true } }, reviews: true },
    });

    if (!booking) {
      throw new NotFoundException("Booking not found.");
    }
    if (booking.userId !== user.userId) {
      throw new ForbiddenException("You can only review your own bookings.");
    }
    if (booking.status !== "COMPLETED") {
      throw new BadRequestException(
        "A stay can only be reviewed once the booking is marked COMPLETED."
      );
    }
    if (booking.reviews.length > 0) {
      throw new ConflictException("This booking has already been reviewed.");
    }

    return this.prisma.review.create({
      data: {
        propertyId: booking.unit.propertyId,
        bookingId: booking.id,
        userId: user.userId,
        overallRating: dto.overallRating,
        categories: {
          create: dto.categories.map((c) => ({
            category: c.category,
            rating: c.rating,
          })),
        },
      },
      include: { categories: true },
    });
  }

  /**
   * A manager (member of the property's organisation) or Admin may respond
   * once, publicly, to a review — mirrors assertCanManageOrganisation's
   * access rule in properties.service.ts.
   */
  async respond(user: AuthenticatedUser, reviewId: string, dto: CreateReviewResponseDto) {
    const review = await this.prisma.review.findUnique({
      where: { id: reviewId },
      include: { property: { select: { organisationId: true } } },
    });
    if (!review) {
      throw new NotFoundException("Review not found.");
    }

    if (!user.roles.includes("Admin")) {
      const membership = await this.prisma.organisationMember.findUnique({
        where: {
          organisationId_userId: { organisationId: review.property.organisationId, userId: user.userId },
        },
      });
      if (!membership) {
        throw new ForbiddenException("You do not manage the property this review is about.");
      }
    }

    return this.prisma.reviewResponse.create({
      data: {
        reviewId,
        responderId: user.userId,
        body: dto.body,
      },
    });
  }

  /** Public, paginated-free list for a property — same data already embedded
   * in getPropertyBySlug, exposed standalone for the account/manager views. */
  async listForProperty(propertyId: string) {
    return this.prisma.review.findMany({
      where: { propertyId },
      include: { categories: true, responses: true },
      orderBy: { createdAt: "desc" },
    });
  }
}

File: apps/api-py/app/services/reviews_service.py
"""Direct port of reviews/reviews.service.ts."""

from fastapi import HTTPException
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.deps import AuthenticatedUser
from app.models import (
    Booking,
    OrganisationMember,
    Property,
    Review,
    ReviewCategory,
    ReviewResponse,
    Unit,
)

REVIEW_CATEGORY_KEYS = [
    "accuracy", "security", "water", "internet", "cleanliness", "management", "value", "distance",
]


async def create_for_booking(
    db: AsyncSession, user: AuthenticatedUser, booking_id: str, overall_rating: float, categories: list[dict]
) -> Review:
    """Only the tenant on a COMPLETED booking may review the stay, and
    only once per booking — the "verified stays only" rule."""
    booking = (await db.execute(select(Booking).where(Booking.id == booking_id))).scalar_one_or_none()
    if booking is None:
        raise HTTPException(status_code=404, detail="Booking not found.")
    if booking.user_id != user.user_id:
        raise HTTPException(status_code=403, detail="You can only review your own bookings.")
    if booking.status != "COMPLETED":
        raise HTTPException(status_code=400, detail="A stay can only be reviewed once the booking is marked COMPLETED.")

    existing_reviews = (await db.execute(select(Review).where(Review.booking_id == booking_id))).scalars().all()
    if len(existing_reviews) > 0:
        raise HTTPException(status_code=409, detail="This booking has already been reviewed.")

    unit = (await db.execute(select(Unit).where(Unit.id == booking.unit_id))).scalar_one()

    review = Review(property_id=unit.property_id, booking_id=booking.id, user_id=user.user_id, overall_rating=overall_rating)
    db.add(review)
    await db.flush()

    for c in categories:
        db.add(ReviewCategory(review_id=review.id, category=c["category"], rating=c["rating"]))

    await db.commit()
    await db.refresh(review)
    return review


async def respond(db: AsyncSession, user: AuthenticatedUser, review_id: str, body: str) -> ReviewResponse:
    """A manager (member of the property's organisation) or Admin may
    respond once, publicly, to a review."""
    review = (await db.execute(select(Review).where(Review.id == review_id))).scalar_one_or_none()
    if review is None:
        raise HTTPException(status_code=404, detail="Review not found.")

    prop = (await db.execute(select(Property).where(Property.id == review.property_id))).scalar_one()

    if "Admin" not in user.roles:
        membership = (
            await db.execute(
                select(OrganisationMember).where(
                    OrganisationMember.organisation_id == prop.organisation_id,
                    OrganisationMember.user_id == user.user_id,
                )
            )
        ).scalar_one_or_none()
        if membership is None:
            raise HTTPException(status_code=403, detail="You do not manage the property this review is about.")

    response = ReviewResponse(review_id=review_id, responder_id=user.user_id, body=body)
    db.add(response)
    await db.commit()
    await db.refresh(response)
    return response


async def list_for_property(db: AsyncSession, property_id: str) -> list[Review]:
    result = await db.execute(select(Review).where(Review.property_id == property_id).order_by(Review.created_at.desc()))
    return list(result.scalars().all())
